ethr/ethr.go

```
@@ -0,0 +1,473 @@
+							}
+							pk.PublicKeyMultibase = publicKeyMultibase
+						case "base58":
+							keyOrServiceBytes, err := hex.DecodeString(keyOrService)
+							if err != nil {
+								return nil, err
+							}
+							publicKeyMultibase, err := multibase.Encode(multibase.Base58BTC, keyOrServiceBytes)
+							if err != nil {
+								return nil, err
+							}
+							pk.PublicKeyMultibase = publicKeyMultibase
+						case "pem":
+							pk.PublicKeyPem = keyOrService
+						default:
+							// TODO: Should we have a default case?
+						}
+						pks.Set(eventIndex, pk)
+						if kind == "sigAuth" {
+							auths.Set(eventIndex, pk.ID)
+						}
+					case "svc":
+						serviceCount++
+						keyOrServiceBytes, err := hex.DecodeString(keyOrService)
+						if err != nil {
+							return nil, err
+						}
+						services.Set(eventIndex, resolver.ServiceEndpoint{
+							ID:              fmt.Sprintf("%s#service-%d", did, serviceCount),
+							Type:            algorithm,
+							ServiceEndpoint: string(keyOrServiceBytes),
+						})
+					}
+				}
+			}
+		} else {
+			var eventIndex string
+			if event.Type == "DIDDelegateChanged" {
+				delegateCount++
+				delegateType := trimmer.ReplaceAllString(string(event.DelegateType[:]), "")
+				eventIndex = fmt.Sprintf("%s-%s-%s", event.Type, delegateType, event.Delegate.Hex())
+			} else if event.Type == "DIDAttributeChanged" {
+				attributeName := trimmer.ReplaceAllString(string(event.Name[:]), "")
+				eventIndex = fmt.Sprintf("%s-%s-%s", event.Type, attributeName, event.Value)
+				if strings.HasPrefix(string(event.Name[:]), "did/pub/") {
+					delegateCount++
+				} else if strings.HasPrefix(string(event.Name[:]), "did/svc/") {
+					serviceCount++
+				}
+			}
+			auths.Delete(eventIndex)
+			pks.Delete(eventIndex)
+			services.Delete(eventIndex)
+			if event.Type == "DIDOwnerChanged" {
+				if event.Owner.Hex() == (common.Address{}).Hex() {
+					deactivated = true
+					// TODO: Should we still build up the document?
```

What does this mean again? Owner DID is no longer good? Maybe return an error?

---

Oh, I see below you are return an error in this case